### ISSUE TITLE
define CF and MR dependencies independently

### ISSUE THREAD

**Issue by DAmNRelentless** [Kir-Antipov/mc-publish#13]
Right now, I use the action for two steps in a row. One for each hosting platform because the dependencies that exist on CurseForge are either not existent on Modrinth or they have different slugs.

Would be cool if you could add `modrinth-dependencies` and `curseforge-dependencies` and that the normal `dependencies` option is used as common option.

**Comment by Kir-Antipov** [comment]
It works this exact way, it's just not documented ;)

The thing about GitHub Actions is you cannot pass complex objects to them as inputs, therefore I use some kind of self-built convention - inputs like `modrinth-id: id`, `modrinth-token: token` will be transformed into `{ "modrinth": { "id": "id", "token": "token" } }` object, which is an uploader-specific options object. Then it is being combined with "common" options object, which contains keys like `dependencies`, `files`, `version`, `name`, and other root entries.

Therefore, nothing really stops you from making a configuration such as this one:
```yaml
files: build/libs/build.jar # Default files value

github-token: {{ secrets.GITHUB_TOKEN }}

modrinth-id: modrinth-id
modrinth-token: {{ secrets.MODRINTH_TOKEN }}
modrinth-files: build/libs/*.jar # Modrinth-specific files, overwrites top entry of `files`
modrinth-dependencies: fabric-api # Modrinth-specific dependencies

curseforge-id: curseforge-id
curseforge-token: {{ secrets.CURSEFORGE_TOKEN }}
curseforge-files: build/libs/!(*-@(dev|sources)).jar # CurseForge-specific files, overwrites top entry of `files`
curseforge-dependencies: cloth-config # CurseForge-specific dependencies
```

I did not document this behavior because GitHub emits warnings for inputs that weren't specified in `action.yml`, and I didn't want to duplicate millions *(ok, not that many, but you've got the point)* of top-level inputs for every available uploader (`modrinth`, `curseforge`, and `github` at the moment), at least manually. I will make a script that does it automatically in the foreseen future, until then you're free to use this functionality and ignore warnings from GitHub.

Leaving this open, so I won't forget to document this feature

**Comment by DAmNRelentless** [comment]
Awesome! Thanks for the fast and detailed response.

**Comment by Kir-Antipov** [comment]
You are welcome :)

**Comment by DAmNRelentless** [comment]
One question that is left:
Even if you have target-specific options, this wouldn't work for multiloader projects without using multiple steps, wouldn't it?

E.g. you have Forge and Fabric in one repo and you want to publish them to CF, MR and GH. The JARs that are built are not universal which means you have a `modXY-fabric-1.18.2-2.0.0.jar` and a `modXY-forge-1.18.2-2.0.0.jar`. Is there any way to upload them with a single step with the action?

**Comment by Kir-Antipov** [comment]
> Even if you have target-specific options, this wouldn't work for multiloader projects without using multiple steps, wouldn't it?

Seems like it, because you need some inputs to be in some kind of superposition, e.g., `loaders` being equal to `["fabric"]` and `["forge"]` at the same time. I do not see a good solution to handle such cases either, because it looks like you need to specify different set of options for each set of files, therefore it would be practical to have two different steps for that. At the end of the day, these are still different mods that just happened to have similar branding ;)

> If one of the previous steps would create the release already, the other step would fail because the GitHub release already exists

No, it should not. At least I do not see why it should, I would consider this behavior a bug. Earlier, `mc-publish` was not able to create GitHub Releases, therefore the `tag` input existed so users could specify an existing release by themselves. And that code did not go anywhere, `mc-publish` tries to create a release if does not already exist, otherwise, it just updates it content in such a way, that new files will be appended to the one, and duplicate files will replace older ones.

**Comment by Kir-Antipov** [comment]
On the other hand, I can finally make use of `files-primary` returning more than one file, so when the `loaders` input is omitted auto-detection can kick in and group files by their supported mod loaders, and then we will have N groups of files which can be uploaded to their targets consequentially

**Comment by Kir-Antipov** [comment]
Yep, I like this idea, and I will definitely implement it. It seem we are getting v4.0 instead of v3.1 :D

**Comment by Kir-Antipov** [comment]
> It would just make the configuration too complex and complicated to understand

Yep, definitely

> Cannot find or create release #undefined

Oh... That's my bad :D I'm checking for `GITHUB_REF` too late. This will be fixed in the upcoming v3.1, so don't throw you configuration out yet ;)
